File: src/tools/pullNotion.ts
#!/usr/bin/env bun
import path from "node:path";
import {
  asArray,
  asObject,
  asObjects,
  asString,
  firstString,
  frontmatter,
  loadDotenv,
  requireString,
  slugify,
  utcNow,
  wikiRoot,
  writeOnce,
  type JsonObject,
} from "./common.js";

const rawDir = path.join(wikiRoot, "raw", "notion");

type Args = {
  pageId: string;
  dryRun: boolean;
};

function parseArgs(argv: string[]): Args {
  const args: Partial<Args> = { dryRun: false };
  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    if (arg === "--help" || arg === "-h") {
      console.log("usage: pullNotion --page-id PAGE_ID [--dry-run]");
      process.exit(0);
    }
    if (arg === "--page-id") {
      args.pageId = requireString(argv[++index], "--page-id requires a value");
    } else if (arg === "--dry-run") {
      args.dryRun = true;
    } else {
      throw new Error(`Unknown argument: ${arg}`);
    }
  }
  if (!args.pageId) {
    throw new Error("--page-id is required");
  }
  return { pageId: args.pageId, dryRun: args.dryRun ?? false };
}

async function notionRequest(apiPath: string, token: string, version: string): Promise<JsonObject> {
  const response = await fetch(`https://api.notion.com/v1${apiPath}`, {
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${token}`,
      "Notion-Version": version,
    },
  });
  if (!response.ok) {
    throw new Error(`HTTP ${response.status} fetching Notion ${apiPath}`);
  }
  const payload = asObject(await response.json());
  if (!payload) {
    throw new Error(`Unexpected Notion response for ${apiPath}`);
  }
  return payload;
}

function richTextText(value: unknown): string {
  return asArray(value)
    .map((item) => firstString(asObject(item) ?? {}, ["plain_text"]))
    .join("");
}

function pageTitle(page: JsonObject): string {
  const properties = asObject(page["properties"]) ?? {};
  for (const property of Object.values(properties)) {
    const object = asObject(property);
    if (object?.["type"] === "title") {
      const title = richTextText(object["title"]);
      if (title) {
        return title;
      }
    }
  }
  return "Untitled Notion page";
}

async function fetchBlocks(blockId: string, token: string, version: string): Promise<JsonObject[]> {
  const blocks: JsonObject[] = [];
  let startCursor: string | undefined;
  do {
    const params = new URLSearchParams({ page_size: "100" });
    if (startCursor) {
      params.set("start_cursor", startCursor);
    }
    const payload = await notionRequest(`/blocks/${encodeURIComponent(blockId)}/children?${params.toString()}`, token, version);
    blocks.push(...asObjects(payload["results"]));
    startCursor = asString(payload["next_cursor"]) || undefined;
    if (!payload["has_more"]) {
      break;
    }
  } while (startCursor);
  return blocks;
}

function blockToMarkdown(block: JsonObject): string[] {
  const blockType = asString(block["type"]);
  const data = asObject(block[blockType]) ?? {};
  const text = richTextText(data["rich_text"]);

  switch (blockType) {
    case "heading_1":
      return [`# ${text}`, ""];
    case "heading_2":
      return [`## ${text}`, ""];
    case "heading_3":
      return [`### ${text}`, ""];
    case "bulleted_list_item":
      return [`- ${text}`];
    case "numbered_list_item":
      return [`1. ${text}`];
    case "to_do": {
      const mark = data["checked"] ? "x" : " ";
      return [`- [${mark}] ${text}`];
    }
    case "quote":
    case "callout":
      return [`> ${text}`, ""];
    case "code": {
      const language = asString(data["language"]);
      return [`\`\`\`${language}`, text, "```", ""];
    }
    case "child_page":
      return [`- Child page: ${firstString(data, ["title"], "Child page")}`];
    case "paragraph":
      return text ? [text, ""] : [""];
    default:
      return [`<!-- Unsupported Notion block: ${blockType || "unknown"} -->`, ""];
  }
}

async function renderBlocks(blocks: JsonObject[], token: string, version: string, depth = 0): Promise<string[]> {
  const lines: string[] = [];
  for (const block of blocks) {
    lines.push(...blockToMarkdown(block));
    if (block["has_children"] && depth < 3) {
      const blockId = asString(block["id"]);
      if (blockId) {
        lines.push(...(await renderBlocks(await fetchBlocks(blockId, token, version), token, version, depth + 1)));
      }
    }
  }
  return lines;
}

async function main(): Promise<number> {
  const args = parseArgs(process.argv.slice(2));
  await loadDotenv();
  const token = process.env["NOTION_TOKEN"] ?? "";
  const version = process.env["NOTION_VERSION"] ?? "2022-06-28";
  if (!token) {
    console.error("Set NOTION_TOKEN in .env.");
    return 2;
  }

  const page = await notionRequest(`/pages/${encodeURIComponent(args.pageId)}`, token, version);
  const blocks = await fetchBlocks(args.pageId, token, version);
  const title = pageTitle(page);
  const editedTime = asString(page["last_edited_time"]);
  const date = editedTime ? editedTime.slice(0, 10) : utcNow().toISOString().slice(0, 10);
  const filePath = path.join(rawDir, `${date}-${slugify(title, "notion-page")}.md`);
  const pageUrl = asString(page["url"]);
  const content =
    frontmatter({
      type: "raw_notion_page",
      source: "notion",
      date,
      title,
      page_id: args.pageId,
      source_url: pageUrl || null,
      pulled_at: utcNow().toISOString(),
    }) + `\n# ${title}\n\n${(await renderBlocks(blocks, token, version)).join("\n").trimEnd()}\n`;

  if (args.dryRun) {
    console.log(path.relative(wikiRoot, filePath));
  } else if (await writeOnce(filePath, content)) {
    console.log(`wrote ${path.relative(wikiRoot, filePath)}`);
  } else {
    console.log(`skipped existing ${path.relative(wikiRoot, filePath)}`);
  }
  return 0;
}

main().then(
  (code) => {
    process.exitCode = code;
  },
  (error: unknown) => {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  },
);

File: tools/pull_notion.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import datetime as dt
import json
import os
import sys
import urllib.parse
import urllib.request
from pathlib import Path
from typing import Any

from wiki_common import WIKI_ROOT, frontmatter, load_dotenv, slugify, utc_now, write_once


RAW_DIR = WIKI_ROOT / "raw" / "notion"


def notion_request(path: str, token: str, version: str) -> dict[str, Any]:
    url = "https://api.notion.com/v1" + path
    req = urllib.request.Request(
        url,
        headers={
            "Authorization": f"Bearer {token}",
            "Notion-Version": version,
            "Accept": "application/json",
        },
    )
    with urllib.request.urlopen(req, timeout=30) as response:
        return json.loads(response.read().decode("utf-8"))


def rich_text_text(value: list[dict[str, Any]] | None) -> str:
    if not value:
        return ""
    return "".join(str(item.get("plain_text") or "") for item in value)


def page_title(page: dict[str, Any]) -> str:
    properties = page.get("properties") or {}
    for prop in properties.values():
        if isinstance(prop, dict) and prop.get("type") == "title":
            title = rich_text_text(prop.get("title"))
            if title:
                return title
    return "Untitled Notion page"


def fetch_blocks(block_id: str, token: str, version: str) -> list[dict[str, Any]]:
    blocks: list[dict[str, Any]] = []
    start_cursor: str | None = None
    while True:
        query = {"page_size": "100"}
        if start_cursor:
            query["start_cursor"] = start_cursor
        payload = notion_request(f"/blocks/{urllib.parse.quote(block_id)}/children?{urllib.parse.urlencode(query)}", token, version)
        blocks.extend(payload.get("results", []))
        if not payload.get("has_more"):
            break
        start_cursor = payload.get("next_cursor")
    return blocks


def block_to_markdown(block: dict[str, Any]) -> list[str]:
    block_type = block.get("type")
    data = block.get(block_type, {}) if isinstance(block_type, str) else {}
    text = rich_text_text(data.get("rich_text"))
    if block_type == "heading_1":
        return [f"# {text}", ""]
    if block_type == "heading_2":
        return [f"## {text}", ""]
    if block_type == "heading_3":
        return [f"### {text}", ""]
    if block_type == "bulleted_list_item":
        return [f"- {text}"]
    if block_type == "numbered_list_item":
        return [f"1. {text}"]
    if block_type == "to_do":
        mark = "x" if data.get("checked") else " "
        return [f"- [{mark}] {text}"]
    if block_type == "quote":
        return [f"> {text}", ""]
    if block_type == "code":
        language = data.get("language") or ""
        return [f"```{language}", text, "```", ""]
    if block_type == "callout":
        return [f"> {text}", ""]
    if block_type == "child_page":
        title = data.get("title") or "Child page"
        return [f"- Child page: {title}"]
    if block_type == "paragraph":
        return [text, ""] if text else [""]
    return [f"<!-- Unsupported Notion block: {block_type} -->", ""]


def render_blocks(blocks: list[dict[str, Any]], token: str, version: str, depth: int = 0) -> list[str]:
    lines: list[str] = []
    for block in blocks:
        lines.extend(block_to_markdown(block))
        if block.get("has_children") and depth < 3:
            children = fetch_blocks(block["id"], token, version)
            child_lines = render_blocks(children, token, version, depth + 1)
            lines.extend(child_lines)
    return lines


def main() -> int:
    parser = argparse.ArgumentParser(description="Snapshot one Notion page into raw/notion.")
    parser.add_argument("--page-id", required=True)
    parser.add_argument("--dry-run", action="store_true")
    args = parser.parse_args()

    load_dotenv()
    token = os.environ.get("NOTION_TOKEN", "")
    version = os.environ.get("NOTION_VERSION", "2022-06-28")
    if not token:
        print("Set NOTION_TOKEN in .env.", file=sys.stderr)
        return 2

    try:
        page = notion_request(f"/pages/{urllib.parse.quote(args.page_id)}", token, version)
        blocks = fetch_blocks(args.page_id, token, version)
    except Exception as exc:
        print(f"Failed to fetch Notion page: {exc}", file=sys.stderr)
        return 1

    title = page_title(page)
    edited_time = str(page.get("last_edited_time") or "")
    date = edited_time[:10] if edited_time else dt.datetime.now(dt.timezone.utc).date().isoformat()
    path = RAW_DIR / f"{date}-{slugify(title, 'notion-page')}.md"
    page_url = str(page.get("url") or "")
    content = (
        frontmatter(
            {
                "type": "raw_notion_page",
                "source": "notion",
                "date": date,
                "title": title,
                "page_id": args.page_id,
                "source_url": page_url or None,
                "pulled_at": utc_now().isoformat(timespec="seconds"),
            }
        )
        + f"\n# {title}\n\n"
        + "\n".join(render_blocks(blocks, token, version)).rstrip()
        + "\n"
    )

    if args.dry_run:
        print(path.relative_to(WIKI_ROOT))
        return 0
    if write_once(path, content):
        print(f"wrote {path.relative_to(WIKI_ROOT)}")
    else:
        print(f"skipped existing {path.relative_to(WIKI_ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
